src/components/VmActions/AutoSelectConsole.js

```diff
@@ -0,0 +1,144 @@
+import React, { useEffect, useState } from 'react'
+import { connect } from 'react-redux'
+import PropTypes from 'prop-types'
+import { addUserMessage, downloadVmConsoles } from '_/actions'
+import ConsoleConfirmationModal from '_/components/VmActions/ConsoleConfirmationModal'
+import { msg } from '_/intl'
+import { isWindows } from '_/helpers'
+import {
+  EMPTY_CONSOLES_LIST,
+  NO_DEFAULT_CONSOLE,
+  DISPATCH_STATUSES,
+} from '_/constants/consoleAutoSelect'
+const {
+  DO_NOT_DISPATCH,
+  SHOULD_DISPATCH,
+  DISPATCHED,
+} = DISPATCH_STATUSES
+const defaultConsoleProtocolErrorsList = [
+  NO_DEFAULT_CONSOLE,
+  EMPTY_CONSOLES_LIST,
+]
+
+const AutoSelectConsole = ({
+  vm,
+  show,
+  defaultConsole,
+  isNoVNC,
+  fetchConsoles,
+  id,
+  onFailure,
+  onClose,
+}) => {
+  const initialState = {
+    selectedProtocol: undefined,
+    failureReason: undefined,
+    dispatchStatus: DO_NOT_DISPATCH,
+  }
+
+  const [state, setState] = useState(initialState)
+  const {
+    dispatchStatus,
+    selectedProtocol,
+  } = state
+
+  const handleFailure = (reason, vmName, hasRdp) => {
+    let message
+    switch (reason) {
+      case EMPTY_CONSOLES_LIST:
+        message = hasRdp
+          ? msg.cantOpenNonRdpGraphicalConsoleEmptyList({ vmName })
+          : msg.cantOpenGraphicalConsoleEmptyList({ vmName })
+        break
+      case NO_DEFAULT_CONSOLE:
+        message = hasRdp
+          ? msg.cantOpenNonRdpGraphicalConsoleNoDefault({ vmName, defaultConsole })
+          : msg.cantOpenGraphicalConsoleNoDefault({ vmName, defaultConsole })
+        break
+      default:
+        message = hasRdp
+          ? msg.cantOpenNonRdpGraphicalConsoleMoreOptions()
+          : msg.cantOpenGraphicalConsoleMoreOptions()
+        break
+    }
+    onFailure(message)
+  }
+
+  const handleClose = () => {
+    setState(state => ({ ...state, dispatchStatus: DO_NOT_DISPATCH }))
+    onClose()
+  }
+
+  useEffect(() => {
+    if (show) {
+      setState(state => ({
+        ...state,
+        dispatchStatus: SHOULD_DISPATCH,
+      }))
+      const vmId = vm.get('id')
+      fetchConsoles(vmId)
+    }
+  }, [show, fetchConsoles])
+
+  useEffect(() => {
+    if (show && dispatchStatus === SHOULD_DISPATCH) {
+      const newState = {}
+      const selectedProtocol = vm.get('defaultConsole')
```

Comment:
`defaultConsole` is used here in 2 contexts: `config.defaultConsole` and `vm.defaultConsole`.
Perhaps we should rename the second (according to its usage) to `selectedProtocol`?

Reply:
IMO use `selectedProtocol` can be confusing because inside `AutoSelectConsole` component it's the selectedConsole but in the context of the vm is the default, it's individual and some people might think different...

I leave it that way unless the rest of the team thinks different...